rustler_mix/lib/rustler/compiler.ex

```diff
@@ -1,58 +1,42 @@
-defmodule Mix.Tasks.Compile.Rustler do
-  @moduledoc """
-  Compiles Erlang NIFs written in Rust by calling into `cargo`.
-
-  This task extracts the configuration for Rustler crates from `mix.exs`
-  and spawns a `cargo` process to do the actual compilation.
-
-  This task is usually used by configuring the compilers in `mix.exs`. In
-  `project/0`, add `compilers: [:rustler] ++ Mix.compilers(),`. Then, `mix compile`
-  will automatically pick up `mix compile.rustler`.
-  """
-
-  @shortdoc "Compiles Erlang NIFs written in Rust by calling into `cargo`"
-
-  use Mix.Task
-  require Logger
+defmodule Rustler.Compiler do
+  @moduledoc false
 
   alias Rustler.Compiler.{Messages, Rustup}
 
-  @recursive true
-
-  def run(_args) do
-    config = Mix.Project.config()
-    crates = Keyword.get(config, :rustler_crates) || raise_missing_crates()
-
-    File.mkdir_p!(priv_dir())
-
-    Enum.map(crates, &compile_crate/1)
-
-    # Workaround for a mix problem. We should REALLY get this fixed properly.
-    _ =
-      symlink_or_copy(
-        config,
-        Path.expand("priv"),
-        Path.join(Mix.Project.app_path(config), "priv")
-      )
-  end
-
-  defp priv_dir, do: "priv/native"
-
-  defp compile_crate({name, config}) do
-    crate_path = Keyword.get(config, :path, "native/#{name}")
-    build_mode = Keyword.get(config, :mode, :release)
-
-    Mix.shell().info("Compiling NIF crate #{inspect(name)} (#{crate_path})...")
+  @doc false
+  def compile_crate(module, opts) do
+    otp_app = Keyword.fetch!(opts, :otp_app)
+    config = Application.get_env(otp_app, module, [])
 
+    crate = to_string(config[:crate] || opts[:crate] || otp_app)
+    load_data = config[:load_data] || opts[:load_data] || 0
+    crate_path = Keyword.get(config, :path, "native/#{crate}")
     crate_full_path = Path.expand(crate_path, File.cwd!())
+    build_mode = config[:mode] || opts[:mode] || build_mode(Mix.env())
+
+    external_resources =
+      "#{crate_full_path}/**/*"
+      |> Path.wildcard()
+      |> Enum.reject(fn path ->
+        String.starts_with?(path, "#{crate_full_path}/target/")
+      end)
 
     target_dir =
       Keyword.get(
         config,
         :target_dir,
-        Path.join([Mix.Project.build_path(), "rustler_crates", Atom.to_string(name)])
+        Path.join([Mix.Project.build_path(), "lib", "#{otp_app}", "native", "#{crate}"])
       )
 
+    priv_dir =
+      otp_app
+      |> :code.priv_dir()
+      |> to_string()
+
+    File.mkdir_p!(priv_dir())
+
+    lib_path = String.to_charlist("#{priv_dir}/native/lib#{crate}")
```

Comment:
Should we decide on a common scheme here for `rebar3_cargo` and this? What I'm doing there (right now) is `#{priv_dir}/crates/#{name}/#{version}/#{mode}/` and drop all build results in there.

Reply:
So...the way this works currently is...`cargo` compiles everything in `_build/#{Mix.env()}/lib/#{otp_app}/native/#{crate}`.

Then only the necessary artifacts are moved into `priv/native`.

Reply:
That will just effectively prevent you from being able to reload, as some OSes won't load a changed library from the same location. Windows will even prevent overwriting the in-use DLL. And to be able to reliably get the build target and version you will need to parse `cargo`'s output (not a huge deal, though, it's very simple, see https://github.com/filmor/rebar3_cargo/blob/cde49c2a783e523090d88a77549228efd81d110f/src/rebar3_cargo_compile_prv.erl#L82-L102).

Reply:
@filmor is this something we want to do here in this PR or can it wait until after we incorporate with `rebar3_cargo`?